netconf_client/ncclient.py

```diff
@@ -21,40 +25,201 @@
+
+            Manager.logger().debug(
+                "NC Request%s:\n%s",
+                conn_id,
+                pretty,
+                extra={"ncclient.Manager.funcname": funcname},
+            )
+
+    def _log_rpc_response(self, rpc_xml, funcname):
+        if self._is_logger_enabled():
+            end_time = self._get_timestamp()
+            conn_id = self._get_connection_info("<=")
+
+            taken = end_time - self._start_time
+            taken_formatted = "%d.%03d" % (taken.seconds, taken.microseconds / 1000)
+            pretty = _pretty_xml(rpc_xml) if rpc_xml else "(None)"
+
+            Manager.logger().debug(
+                "NC Response%s (%s sec):\n%s",
+                conn_id,
+                taken_formatted,
+                pretty,
+                extra={"ncclient.Manager.funcname": funcname},
+            )
+
+    def _log_rpc_failure(self, message, funcname):
+        if self._is_logger_enabled():
+            end_time = self._get_timestamp()
+            conn_id = self._get_connection_info("<=")
+
+            taken = end_time - self._start_time
+            taken_formatted = "%d.%03d" % (taken.seconds, taken.microseconds / 1000)
+            if message:
+                message = "Cause: {}\n".format(message)
+
+            Manager.logger().debug(
+                "NC Failure %s (%s sec)\n%s",
+                conn_id,
+                taken_formatted,
+                message,
+                extra={"ncclient.Manager.funcname": funcname},
+            )
+
+    def _send_rpc(self, rpc_xml):
+        """Send given NC request message and expect a NC response
+
+           Both, the NC request and response messages are logged with timestamp.
+           In case of failure or exceptions, the error cause is logged, if known.
+           Exceptions thrown by functions called by _send_rpc() are re-raised
+           after they have been logged.
+
+           :param str rpc_xml: XML RPC message to sent to NC server
+
+           :rtype :tupel: (`str` raw XML response, `ElementTree`: Element Tree or None)
+           :exception: whatever exceptions raised by /netconf-client/netconf_client/ncclient.py
+        """
+
+        (raw, ele) = (None, None)
+        funcname = inspect.stack()[1][3]
+        self._log_rpc_request(rpc_xml, funcname)
+
+        try:
+            f = self.session.send_rpc(rpc_xml)
+            r = f.result(timeout=self.timeout)
+            if not r:
+                self._log_rpc_failure("RPC returned without result", funcname)
+            else:
+                (raw, ele) = f.result(timeout=self.timeout)
+                self._log_rpc_response(raw, funcname)
+        except CancelledError:
+            self._log_rpc_failure("RPC cancelled", funcname)
+            raise
+        except TimeoutError:
+            self._log_rpc_failure(
+                "RPC timeout (max. {} seconds)".format(self.timeout), funcname
+            )
+            raise
+        except Exception as e:
```

Comment:
Ah yeah I forgot to mention in the last round, this is exactly the right use case for catching all Exceptions. :+1: